### ISSUE TITLE
DX12: Memory leaks with intel gpu & 30+ driver

### ISSUE THREAD

**Issue by xiaopengli89** [gfx-rs/wgpu#6143]
**Description**
DX12 backend has memory leaks on some Intel drivers.

**Repro steps**

1. Intel integrated gpu
2. Setup intel 30+ gpu driver, e.g. 31.0.101.2128
3. Running water example with dx12 backend

**Extra materials**
<img width="1762" alt="image" src="https://github.com/user-attachments/assets/db1732d6-b7b7-4e2f-87a0-e35e0924a7f9">

![20240822-163346](https://github.com/user-attachments/assets/7dceb2b3-a18c-4823-a42a-49b5664c71c4)

**Platform**
Windows 10, wgpu 22.0

**Comment by teoxoy** [comment]
Is this a new leak?

**Comment by xiaopengli89** [comment]
> Is this a new leak?

No, it can be reproduced on older versions of wgpu, but there is no leak on the 27.x.x.x driver, maybe it's a driver bug.

**Comment by jimblandy** [comment]
I understand that the water example is a reproducer, so that's in principle all we need to look into this. But if you're able to, it would be nice to get a smaller test case, since water is one of the most complex of the examples.

**Comment by VadikPtr** [comment]
I have the same problem, but I'm using wgpu-native.  
Specs:
```
RTX 3090 TI
WGPU version: 22.1.0.5
Backend: Vulkan, DX12 (doesn't depend on particular backend)
```
It looks like `wgpu-core\src\device\queue.rs::PendingWrites::temp_resources` was not released somewhere. I don't know rust and can't tell the reason. 

Allocation stack trace:
![image](https://github.com/user-attachments/assets/00afb229-7dc0-43b4-bdf5-f28527e82f51)

**Comment by VadikPtr** [comment]
I made reproduction example in rust https://github.com/cprkv/wgpu_mem_leak (in wgpu-native I do same things, so it has no difference, situation reproduces).

My guess is that uniform buffer is being used by GPU when I write to it. I took this approach from wgpu examples, and there is no comment if it does something wrong. I don't know if there is a correct example of how to do `queue_write_buffer` right. 

I have to add this after some experiments:  
It happens to both backends (Vulkan and DX12) in wgpu-native, but for some reason everything is fine for DX12 in wgpu (rust). Only Vulkan backend has noticeable leakage. I also tried using triple buffering technique (create 3 buffers and use it in round-robin manner), it doesn't help.

**Comment by VadikPtr** [comment]
P.S.: when I updated to version 23.0.1 problem is gone (https://github.com/cprkv/wgpu_mem_leak/tree/problem-solved).

**Comment by cwfitzgerald** [comment]
Going to close this then. If the issue re-occurs please file a new issue!
